port from Python/`py-school-match` to Rust/`matchmaker`

cfm opened the pull request:
https://github.com/cfm/proxy-solver-api/blob/3fe7b27c47830d3cccb1a2c59c9b0e1cf10a0442/README.md?plain=1#L5-L21

cfm committed:
build: cargo new proxy_solver_api --bin

cfm committed:
feat(/health/ready): reimplement health-check endpoint in Rust/Rocket

cfm committed:
test(/health/ready): returns HTTP 204

cfm committed:
feat(/solution): count members and members present

cfm committed:
test(/solution): counts members and members present

cfm committed:
chore(deps): update rocket to 0.5.0-rc.3

cfm committed:
feat(/solution): scaffold stable-marriage problem via matchmaker::{Category, Student}

cfm committed:
build: lint, test

cfm committed:
refactor(/solution): store members present and absent, not all members

Since the solver types these distinctly, they are necessarily disjoint
sets, so there's no point in storing an unused superset of all members.

cfm committed:
feat(/solution): match members absent (Students) to present (Categories)

cfm committed:
feat(/solution): match members only within their preferences

If absent member X does not list member Y in their preferences, member Y
cannot represent member X.

cfm committed:
feat(/solution): return matches rather than counts

cfm committed:
test(big): test real-world attendance snapshot and proxy preferences

Identifiers have been anonymized via the "bin/anonymize" script.

cfm committed:
feat(/solution): use sortable Index{Map,Set} and seeded RNG for reproducibility

cfm committed:
feat(/solution): parameterize solution(capacity)

cfm committed:
build(check): strict "cargo fmt" and "cargo clippy"

cfm committed:
ci: add "check" and "test" jobs

cfm committed:
build(test): ignore "clear" failure on non-interactive terminal (without $TERM)

cfm committed:
build: Rust

cfm committed:
build: add entry-point script to handle environment mapping

cfm committed:
deploy: wrap Rocket in nginx reverse proxy

cfm committed:
deploy: open CORS

https://enable-cors.org/server_nginx.html

cfm committed:
chore: promote Rust over Python implementation

cfm committed:
build(build,run): wrap "docker {build,run}"

cfm committed:
build(venv,check-py): wrap Python setup and linting

cfm committed:
style: Clippy

cfm committed:
feat: log ProxyMetrics struct before and after solving

cfm committed:
refactor(ProxyMetrics): derive SerdeValue for logging

cfm committed:
refactor: ProxyProblem in, ProxySolution out, ProxyMetrics along the way

cfm committed:
refactor: crate::{member,problem,solution}

cfm committed:
test(test_big): load fixtures with include_str!()

Inspired by @legoktm's feedback in
<https://github.com/freedomofpress/securedrop/pull/6925#discussion_r1302562743>.

cfm committed:
feat(/openapi.json): generate OpenAPI specification via rocket_okapi

cfm committed:
docs: readme

cfm committed:
ci: CodeQL doesn't support Rust; remove it

cfm committed:
docs: bin/anonymize

cfm committed:
build(docs): wrap "cargo doc"

cfm committed:
docs: crate

cfm committed:
build(readme): generate via cargo-readme

cfm committed:
refactor(Member::from_info): combine filter()/map() into filter_map().

Co-authored-by: Kunal Mehta <legoktm@debian.org>

cfm committed:
build(check-readme): error if generated readme is out of date

cfm committed:
ci(check): check that generated readme is up to date

cfm committed:
docs(readme): link to architecture

legoktm commented on Makefile:
```diff
@@ -0,0 +1,48 @@
+
+check: check-readme
+	cargo fmt -- --check
+	cargo clippy -- --deny clippy::all
```
You can put this deny inside the code itself, like https://github.com/freedomofpress/securedrop/blob/develop/redwood/src/lib.rs#L1 which helps IDEs, etc. always run clippy with those settings.

legoktm replied:
Also if you enable rustdoc lints (see below), then you should probably also add that to this check target.

legoktm commented on Makefile:
```diff
@@ -0,0 +1,48 @@
+	isort --check $<
+
+check-readme: README.md
+	cargo readme > $<
```
Interesting tool, Rust has [built-in support](https://github.com/rust-lang/rust/blob/master/RELEASES.md#language-18) for going the other way though, writing a README.md and having it included in the rustdoc output.

legoktm commented on entrypoint.sh:
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+/usr/local/bin/proxy-solver-api &
```
Why the `&`?

cfm replied:
Looking at (e.g.) nginxinc/docker-nginx#752, I guess I'm cheating slightly here by using the `nginx` image as both application host and reverse proxy.  The latter expects to run in the foreground, so it doesn't even get run if the `ENTRYPOINT` scripts don't return.

I have to revisit the deployment strategy here anyway, so I'll leave this for now.

legoktm commented on src/main.rs:
```diff
@@ -0,0 +1,164 @@
+#![warn(missing_docs)]
```
Also consider `#![deny(rustdoc::all)]` (example: https://gitlab.wikimedia.org/repos/mwbot-rs/mwbot/-/commit/089d165c246d9e89e0f70da4fd489136792b164e)

legoktm commented on src/main.rs:
```diff
@@ -0,0 +1,164 @@
+
+#[macro_use]
+extern crate derive_new;
+extern crate matchmaker;
```
`extern crate foo;` is only needed if you want to pair it with `#[macro_use]` to import non-namespaced macros. So it's not needed for matchmaker here.

legoktm commented on src/member.rs:
```diff
@@ -0,0 +1,57 @@
+            exclude: present
+                .values()
+                .filter(|&v| !info.preferences.contains(&v.name))
+                .map(|v| present.get(&v.name).unwrap())
```
I would suggest adding a comment like `// unwrap: <explanation of why this won't panic>` with every unwrap(), or using `.expect("explanation...")`

legoktm commented on src/member.rs:
```diff
@@ -0,0 +1,57 @@
+    ///
+    /// Furthermore, since a member CANNOT be represented by an absent member,
+    /// we only store preferences corresponding to present members.
+    fn from_info(info: &MemberInfo, present: &IndexMap<MemberId, Category>) -> Student;
```
`from_*` indicates a conversion function, i.e. it should take ownership of MemberInfo and transform it into something else. Since it doesn't, a better name would be `new_from_info`.

(Unfortunately https://rust-lang.github.io/api-guidelines/naming.html is pretty light on this specific one)

legoktm commented on src/member.rs:
```diff
@@ -0,0 +1,57 @@
+pub type Proxy = Category;
+
+/// A `matchmaker::Student`.
+pub trait Member {
```
I'm curious why you implemented this as a trait instead of just a `create_student()` function or something.

cfm replied:
Probably just a mistakenly object-oriented desire to alias `s/Student/Member/` and `s/Category/Proxy/` everywhere for clarity.  (@tellurideassociation does indeed deal with students, but this tool does not, so seeing `Student` all over the code, rather than a more abstract or theoretical term, is unnecessarily jarring.)

I think I'll leave it this way here.  But I appreciate the reality-check and will keep it in mind.

legoktm commented on src/solution.rs:
```diff
@@ -0,0 +1,142 @@
+
+    /// Gets [`ProxyMetrics`] of the solution *in its current state* (whether
+    /// solved or unsolved).
+    pub fn get_metrics(&mut self) -> ProxyMetrics {
```
Why does this need `&mut self` and not just `&self`?

Also you can just call this `metrics()`, see https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

legoktm commented on src/main.rs:
```diff
@@ -0,0 +1,164 @@
+/// [`solution::ProxySolution`].  `ProxySolution` is deterministic and constant
+/// for a given `ProxyProblem`.
+fn post_solution(log: Slogger, problem: Json<ProxyProblem>) -> Json<ProxySolution> {
+    let mut solution = ProxySolution::from_problem(&problem);
```
ProxySolution ends up being both a container of state and a functional do-er, which I think ends up causing a bit more cloning than necessary. (The other things hinting at needing a bit of restructuring are the `serde(skip)` and `new(default)` helpers)

But I'd probably structure this as something like:

```rust
info!(problem.metrics());
let solution = solve(problem);
info!(solution.metrics());
Json(solution)
```

You can still have `solve()` create a mutable ProxySolution as the first step, but now that's an internal detail and can be refactored later to move stuff around instead of cloning.

cfm replied:
Yeah, this started to smell to me too.  I've done your suggested first-pass refactoring in f81d1cac4d9a2113e08282c9c4f3794cddf4716d and filed #10 for the rest.

legoktm commented on src/main.rs:
```diff
@@ -0,0 +1,164 @@
+
+#[cfg(test)]
+mod test {
+    use super::rocket;
```
I usually just write `use super::*;` to keep everything simpler.

legoktm commented on src/solution.rs:
```diff
@@ -0,0 +1,142 @@
+    fn load_preferences(&mut self, members: &Vec<MemberInfo>) {
+        for info in members {
+            if !self.members_present.contains_key(&info.id) {
+                let absent = <Student as Member>::from_info(info, &self.members_present);
```
You don't need the `as Member` syntax, just `Student::from_info` should work fine. (It's only needed if like two traits define the same function)

legoktm commented on src/solution.rs:
```diff
@@ -0,0 +1,142 @@
+        // Flatten each [`Proxy`]) of `P` → `{A1, A2, ...}` into `A1` → `P`,
+        // `A2` → `P`, ... .
+        for present in self.members_present.values() {
+            for absent in result.placed.get(&present.name).unwrap_or(&Vec::new()) {
```
This feels like a very Python way to do it :) 

Consider avoiding the unwrap with:
```rust
            if let Some(list) = result.placed.get(&present.name) {
                for absent in list {
                    ...
                }
            }
```

legoktm commented on src/member.rs:
```diff
@@ -0,0 +1,57 @@
+
+use matchmaker::{Category, Student};
+
+use rocket::serde::Deserialize;
```
This is a bit of a confusing trap that rocket contains...rocket wanted to re-export all the common dependencies for people to use, but the way macros work is a bit weird...in short, serde's derive macros only work if you have `serde` as a root dependency in your project, or you specify where serde is with the `#[serde(crate = "...")]` attribute (see https://serde.rs/container-attrs.html#crate). 

So despite having `use rocket::serde::Deserialize;`, when you use `#[derive(Deserialize)]`, it's half using the macro from your explicit serde dependency and the other half from rocket's serde dependency. Of course both point to the same thing (otherwise it would blow up pretty badly) but I find it all unnecessarily confusing and I just do `use serde::Deserialize` always.

legoktm reviewed (commented):
Cool project! I hope this feedback is useful :)

cfm commented:
Thank you, @legoktm!  All very useful feedback, especially to make my Pythonisms more Rustlike, that I'll implement soon.

cfm committed:
build(clippy): consolidate settings

cfm committed:
docs: let rustdoc include readme

thread: https://github.com/cfm/proxy-solver-api/pull/9#discussion_r1312824375

cfm committed:
build(check): add rustdoc linter

thread: https://github.com/cfm/proxy-solver-api/pull/9#discussion_r1312828735

cfm committed:
refactor: less Python, more Rust

thread: https://github.com/cfm/proxy-solver-api/pull/9#pullrequestreview-1606617479

cfm committed:
refactor: use serde::Deserialize everywhere

thread: https://github.com/cfm/proxy-solver-api/pull/9#discussion_r1312893812

cfm committed:
refactor: problem::solve() wraps stateful solution::ProxySolution

thread: https://github.com/cfm/proxy-solver-api/pull/9#discussion_r1312868556

cfm committed:
docs: more-concise references
